analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs

```diff
@@ -56,19 +60,31 @@ private static ProgramState LearnBranchingConstraint<T>(ProgramState state, IBin
         {
             return null;
         }
+    }
 
-        ISymbol OperandSymbolWithoutConstraint(IOperation candidate) =>
-            candidate.TrackedSymbol() is { } symbol
-            && (state[symbol] is null || !state[symbol].HasConstraint<T>())
-                ? symbol
-                : null;
-
-        SymbolicConstraint OperandConstraint(IOperation candidate) =>
-            state[candidate] is { } value && value.HasConstraint<T>()
-                ? value.Constraint<T>()
-                : null;
+    private static ProgramState LearnBranchingRelationalConstraint(ProgramState state, IBinaryOperationWrapper binary, bool falseBranch)
+    {
+        // We can fall through ?? because "constraint" and "testedSymbol" are exclusive. Symbols with the constraint will be recognized as "constraint" side.
+        // We only learn in the true branch because not being >, >=, <, <= than a non-empty nullable means either being null or non-null with non-matching value.
+        if (!falseBranch &&
+            (OperandConstraint<ObjectConstraint>(state, binary.LeftOperand) ?? OperandConstraint<ObjectConstraint>(state, binary.RightOperand)) == ObjectConstraint.NotNull
+            && (OperandSymbolWithoutConstraint<ObjectConstraint>(state, binary.LeftOperand) ?? OperandSymbolWithoutConstraint<ObjectConstraint>(state, binary.RightOperand)) is { } testedSymbol
+            && testedSymbol.GetSymbolType().IsNullableValueType())
+        {
+            return state.SetSymbolConstraint(testedSymbol, ObjectConstraint.NotNull);
+        }
+        else
+        {
+            return null;
+        }
     }
```

Comment:
```suggestion
    private static ProgramState LearnBranchingRelationalConstraint(ProgramState state, IBinaryOperationWrapper binary, bool falseBranch) =>
         // We can fall through ?? because "constraint" and "testedSymbol" are exclusive. Symbols with the constraint will be recognized as "constraint" side.
        // We only learn in the true branch because not being >, >=, <, <= than a non-empty nullable means either being null or non-null with non-matching value.
        !falseBranch
        && (OperandConstraint<ObjectConstraint>(state, binary.LeftOperand) ?? OperandConstraint<ObjectConstraint>(state, binary.RightOperand)) == ObjectConstraint.NotNull
        && (OperandSymbolWithoutConstraint<ObjectConstraint>(state, binary.LeftOperand) ?? OperandSymbolWithoutConstraint<ObjectConstraint>(state, binary.RightOperand)) is { } testedSymbol
        && testedSymbol.GetSymbolType().IsNullableValueType()
        ? state.SetSymbolConstraint(testedSymbol, ObjectConstraint.NotNull)
        : null;
```